### PR TITLE
Fixes to allow object-less callables throughout Godot

### LINKED ISSUE #81887: Object-less callables rejected by `Object::connect` et al
### Godot version

master, e3e2528ba7f6e85ac167d687dd6312b35f558591

### System information

linux

### Issue description

Creating this issue to keep track of something discovered while working working on #79005.

Basically, in some places Godot has checks for `object` as a way to check if a callable is valid. This seems wrong, especially for custom callables, which may not actually have an associated object. While this doesn't seem to be a problem within the engine itself, it can be a problem for extensions. In my extension I had to use a [workaround](https://github.com/maiself/godot-python-extension/blob/master/src/variant/callable.cpp#L12) that creates an object that exists for no reason other than to pass those checks.

Here are some examples of object-less callables being rejected:

- [`Object::emit_signalp`](https://github.com/godotengine/godot/blob/e3e2528ba7f6e85ac167d687dd6312b35f558591/core/object/object.cpp#L1100)
- [`Object::connect`](https://github.com/godotengine/godot/blob/e3e2528ba7f6e85ac167d687dd6312b35f558591/core/object/object.cpp#L1303)
- [`Object::_disconnect`](https://github.com/godotengine/godot/blob/e3e2528ba7f6e85ac167d687dd6312b35f558591/core/object/object.cpp#L1388)

I haven't looked to see if there are more cases (I merely searched `object.cpp` for `get_object()`). 

Original comments from #79005:
- https://github.com/godotengine/godot/pull/79005#issuecomment-1724109784
- https://github.com/godotengine/godot/pull/79005#issuecomment-1724579955

### Steps to reproduce

N/A

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by maiself**
This fixes #81887

I searched the entire Godot source code for `Callable` methods `get_object()`, `get_object_id()` and `get_method()`, which could potentially be used to detect if a callable is associated with an object. Several places in the code used these methods to check callable validity instead of using `Callable::is_valid()`, causing object-less callables to be rejected. I then carefully and selectively updated the code to use the correct methods.

A few places remain where object-less callables may not be handled correctly. These files use `get_object()` and `get_method()` for various purposes:

- `scene/main/node.cpp`: serialization
- `editor/editor_node.cpp`: serialization
- `scene/resources/packed_scene.cpp`: serialization
- `core/core_bind.cpp`: error messages
- `core/variant/callable.cpp`: custom callable `get_method()`
- `core/variant/variant.cpp`: error messages

While I was quite thorough, its possible I may have missed things. I also left gdscript and mono alone.

### CODE REVIEW

File: core/object/object.cpp [hunk 1]
```diff
@@ -1390,9 +1391,6 @@ void Object::disconnect(const StringName &p_signal, const Callable &p_callable)
 bool Object::_disconnect(const StringName &p_signal, const Callable &p_callable, bool p_force) {
 	ERR_FAIL_COND_V_MSG(p_callable.is_null(), false, "Cannot disconnect from '" + p_signal + "': the provided callable is null.");
```
Comment (dsnopek) [inline core/object/object.cpp:1392]: Does this line need to be adjusted too? I suspect custom callables will `true` for `is_null()` (which could probably also be changed)

File: core/object/object.cpp [hunk 2]
```diff
@@ -1303,10 +1303,7 @@ void Object::get_signals_connected_to_this(List<Connection> *p_connections) cons
 }
 
 Error Object::connect(const StringName &p_signal, const Callable &p_callable, uint32_t p_flags) {
-	ERR_FAIL_COND_V_MSG(p_callable.is_null(), ERR_INVALID_PARAMETER, "Cannot connect to '" + p_signal + "': the provided callable is null.");
-
-	Object *target_object = p_callable.get_object();
-	ERR_FAIL_NULL_V_MSG(target_object, ERR_INVALID_PARAMETER, "Cannot connect to '" + p_signal + "' to callable '" + p_callable + "': the callable object is null.");
+	ERR_FAIL_COND_V_MSG(!p_callable.is_valid(), ERR_INVALID_PARAMETER, "Cannot connect to '" + p_signal + "': the provided callable is not valid: " + p_callable);
```
Comment (maiself) [inline core/object/object.cpp:1306]: `is_valid()` may be too strong here. If you look at the changes in `editor/plugins` you can see there are callables being created in a constructor that refer to `this`. It seems during the constructor the object isn't fully registered, meaning these callables aren't yet valid, and will be rejected by this line here.

I'm not sure how I like connecting invalid callables, but it was previous behavior to do so, so I'll probably change this line back to use `is_null()`, rather than try to chase down all the not-yet-valid-callables in Godot.

File: editor/plugins/lightmap_gi_editor_plugin.cpp [hunk 3]
```diff
@@ -174,7 +174,7 @@ LightmapGIEditorPlugin::LightmapGIEditorPlugin() {
 	bake->set_icon(EditorNode::get_singleton()->get_editor_theme()->get_icon(SNAME("Bake"), EditorStringName(EditorIcons)));
 	bake->set_text(TTR("Bake Lightmaps"));
 	bake->hide();
-	bake->connect("pressed", Callable(this, "_bake"));
+	bake->connect("pressed", callable_mp(this, &LightmapGIEditorPlugin::_bake));
```
Comment (dsnopek) [inline editor/plugins/lightmap_gi_editor_plugin.cpp:177]: If this `connect()` is the only reason that `"_bake"` is bound in `_bind_methods()`, then that could be removed too.

But I'm assuming this change isn't actually required, right? While using `callable_mp()` is certainly more idiomatic, this is unrelated to the main focus of this PR.

File: editor/plugins/occluder_instance_3d_editor_plugin.cpp [hunk 4]
```diff
@@ -108,7 +108,7 @@ OccluderInstance3DEditorPlugin::OccluderInstance3DEditorPlugin() {
 	bake->set_icon(EditorNode::get_singleton()->get_editor_theme()->get_icon(SNAME("Bake"), EditorStringName(EditorIcons)));
 	bake->set_text(TTR("Bake Occluders"));
 	bake->hide();
-	bake->connect("pressed", Callable(this, "_bake"));
+	bake->connect("pressed", callable_mp(this, &OccluderInstance3DEditorPlugin::_bake));
```
Comment (dsnopek) [inline editor/plugins/occluder_instance_3d_editor_plugin.cpp:111]: Same note here

**Review by dsnopek — COMMENTED** [review]
Thanks!

I went through each change, and tried to think through what it was doing before, and check if the new code is roughly equivalent. For the most part this is looking good to me! I haven't had a chance to do any testing.

File: core/object/object.cpp (hunk 1 above)
Comment (maiself) [reply core/object/object.cpp:1392]: I think custom callables will never return `true` from `is_null()`. My understanding is that `is_null` is a completely empty callable (which do to thier nature `CallableCustom`s can never be), `is_valid` is a non-null callable in a state currently callable.

There was a reason I didn't change this line... trying to remember...

Ah, the reason is a callable may become invalid, but it may still be necessary to disconnect it, so allowing non-null but newly invalid callables here satisfies that.

File: editor/plugins/lightmap_gi_editor_plugin.cpp (hunk 3 above)
Comment (maiself) [reply editor/plugins/lightmap_gi_editor_plugin.cpp:177]: Right

This was a last minute change before submitting the PR. Errors were printing durring editor startup about not being able to connect `"pressed"` to a null object. See my comment on [`Object::connect` line 1306](https://github.com/godotengine/godot/pull/82695/files#r1343282910). I suspect there may be many of these, I only just noticed the two here as they triggered errors on startup.

File: editor/plugins/occluder_instance_3d_editor_plugin.cpp (hunk 4 above)
Comment (maiself) [reply editor/plugins/occluder_instance_3d_editor_plugin.cpp:111]: Since this is exactly the same as the other I'm going to mark it as resolved to keep review clean.

File: core/object/object.cpp [hunk 5]
```diff
@@ -1344,14 +1341,18 @@ Error Object::connect(const StringName &p_signal, const Callable &p_callable, ui
...
+
 	SignalData::Slot slot;
 
 	Connection conn;
 	conn.callable = target;
 	conn.signal = ::Signal(this, p_signal);
 	conn.flags = p_flags;
 	slot.conn = conn;
-	slot.cE = target_object->connections.push_back(conn);
+	if (target_object) {
```
Comment (RandomShaper) [inline core/object/object.cpp:1353]: I was about to ask about how changes to this function will interact with callables of static functions. But I think this indicates that you have them in mind, haven't you?

File: scene/animation/tween.cpp [hunk 6]
```diff
@@ -675,7 +675,7 @@ bool CallbackTweener::step(double &r_delta) {
 		return false;
 	}
 
-	if (!callback.get_object()) {
+	if (!callback.is_valid()) {
```
Comment (KoBeWi) [inline scene/animation/tween.cpp:678]: This breaks static methods. For whatever reason their `is_valid()` returns false, but their Object is the source GDScript. Example:
```GDScript
@tool
extends EditorScript

class CustomNode extends Node:
	static func test():
		print(1)

func _run():
	var tw = Engine.get_main_loop().create_tween()
	tw.tween_callback(CustomNode.test)
```
Does not work with this PR.

Comment (RandomShaper) [reply scene/animation/tween.cpp:678]: Oh, I meant C++ static callables (`callable_mp_static()`). Regarding GDScript, there's #79521, that I think has to do.

Comment (KoBeWi) [reply scene/animation/tween.cpp:678]: Well that issue has to be resolved before these changes.

Comment (dsnopek) [reply scene/animation/tween.cpp:678]: Hrm. This sounds like there's probably a problem with `is_valid()` for static GDScript methods.

The aim of this PR is to fix situations where `callable_mp_static()` is not accepted, which has the exact opposite problem: `get_object()` returns null, but `is_valid()` is true. I think we need to try and standardize on `is_valid()` being the canonical way to check if a callable is valid.

File: core/object/object.cpp (hunk 5 above)
Comment (dsnopek) [reply core/object/object.cpp:1353]: Yeah, the aim of this PR is to fix situations where callables created via `callable_mp_static()` are not accepted (because `get_object()` returns null), but that technically _should_ be able to work.

File: core/object/object.cpp (hunk 1 above)
Comment (dsnopek) [reply core/object/object.cpp:1392]: > I think custom callables will never return `true` from `is_null()`. My understanding is that `is_null` is a completely empty callable

Yeah, I think that's the desired behavior, but if you look at the implementation of `Callable::is_null()`:

```c++
	_FORCE_INLINE_ bool is_null() const {
		return method == StringName() && object == 0;
	}
```

I think a callable created via `callable_mp_static()` would have an empty `method` and `object`, and so return `true` from `is_null()`. Although, I haven't tested it yet, I'm just reading through the code.

Comment (dsnopek) [reply core/object/object.cpp:1392]: Ack! Thanks for pointing that out, I totally didn't notice the `union` :-)

Alright, marking this one as as resolved!

File: core/object/object.cpp (hunk 2 above)
Comment (dsnopek) [reply core/object/object.cpp:1306]: > so I'll probably change this line back to use `is_null()`

That makes sense to me! And then the two changes to use `callable_mp()` where we weren't before can be removed.

File: scene/animation/tween.cpp (hunk 6 above)
Comment (maiself) [reply scene/animation/tween.cpp:678]: I've created #82755 to address the problem with GDScript static methods.

File: core/object/object.cpp (hunk 2 above)
Comment (maiself) [reply core/object/object.cpp]: Just updated the PR for this. I opted for the hybrid approach I mentioned in our discussion on RocketChat, as its actually the most minimal change here (perseveres all original checks). Also removed the `callable_mp()` changes.

File: editor/plugins/lightmap_gi_editor_plugin.cpp (hunk 3 above)
Comment (maiself) [reply editor/plugins/lightmap_gi_editor_plugin.cpp]: I am concerned about these classes not being registered property, but I think that's for a different PR. This PR as it is now should fine for any such classes.

**Comment by maiself** [comment]
Updated as discussed and rebased on master.

File: core/object/object.cpp (hunk 2 above)
Comment (dsnopek) [reply core/object/object.cpp]: Thanks! This looks good to me :-)

**Review by dsnopek — APPROVED** [review]
This is looking good to me personally!

It doesn't fully fix all cases in the code base, but it does fix many of the most common ones, and it does so in a fairly safe and conservative way. When development on Godot 4.3 opens, we can look at the remaining tricky cases in order to fix this comprehensively, hopefully, standardizing on `Callable::is_valid()` everywhere.

But this PR would be nice to get in Godot 4.2 for GDExtension, since we added the ability to create custom callables from GDExtension in PR https://github.com/godotengine/godot/pull/79005, and fixing this issue even somewhat opens up what GDExtensions are able to do.

**Comment by RandomShaper** [comment]
I need a sanity recap about this. Let me rubberduck the situation:

We have the possibility of a `Callable` referring to a static function in C++ or GDScript. In the former case, `CallableCustomStaticMethodPointer` is used (null target object), whereas in the latter, it's a regular `Callable` whose target object is a `Script`.

Assuming the above is correct, are we sure these changes are fine for both cases? I'd say so, since after all the original issue was about the C++ side (wasn't it?) and GDScript static callables would pass the target-must-be-non-null-and-not-dangling check either if done at each call site (before this PR) or implicitly inside the default `is_valid()` implementation.

Is all this disquisition true?

File: core/object/undo_redo.cpp [hunk 7]
```diff
@@ -161,18 +166,23 @@ void UndoRedo::add_undo_method(const Callable &p_callable) {
...
+	ObjectID object_id = p_callable.get_object_id();
+	Object *object = ObjectDB::get_instance(object_id);
+	ERR_FAIL_COND(object_id.is_valid() && object == nullptr);
 
 	Operation undo_op;
 	undo_op.callable = p_callable;
-	undo_op.object = p_callable.get_object_id();
-	if (Object::cast_to<RefCounted>(object)) {
+	undo_op.object = object_id;
+	if (object && Object::cast_to<RefCounted>(object)) {
```
Comment (RandomShaper) [inline core/object/undo_redo.cpp:176]: I've been investigating the performance of `cast_to()` in the case of null objects. My suggestion would be to avoid doing the null check explicitly at the call site and let `cast_to()` do it if desired (it will be inlined). I'm making a PR precisely adding the check so it will be redundant here.

Comment (RandomShaper) [reply core/object/undo_redo.cpp:176]: Here it is: #82903.

**Comment by dsnopek** [comment]
> We have the possibility of a `Callable` referring to a static function in C++ or GDScript. In the former case, `CallableCustomStaticMethodPointer` is used (null target object), whereas in the latter, it's a regular `Callable` whose target object is a `Script`.
>
> Assuming the above is correct, are we sure these changes are fine for both cases?

Yesterday, I re-tested the `EditorScript` that @KoBeWi posted with this PR, to check that GDScript static methods work. I haven't re-tested `callable_mp_static()` in a little while, but that didn't work before and is what this PR aims to fix, so I presume that @maiself has tested it recently. In a couple hours I'll have some more time and I could re-test that too?

**Comment by dsnopek** [comment]
I just tested connecting to a signal with a `callable_mp_static()` using this PR and it worked! Specifically, it was a `Button` and I connected to it's `"pressed"` signal, which was emitted correctly, printing out a test message. So, this PR appears to be working as intended.

**Comment by RandomShaper** [comment]
Thanks for testing it, @dsnopek. (And greetings!) My only pending thing is https://github.com/godotengine/godot/pull/82695#discussion_r1348412393. Otherwise (even if without applying it), this has my blessings.

**Review by RandomShaper — APPROVED** [review]

**Comment by maiself** [comment]
> Assuming the above is correct, are we sure these changes are fine for both cases?

Yep, that's the intent. I've been testing with @KoBeWi's `EditorScript` example, an extended version of the scripts #79521, and my Python bindings, so a lot of types and uses of callables have been tested.

File: core/object/undo_redo.cpp [hunk 8]
```diff
@@ -136,17 +136,22 @@ void UndoRedo::add_do_method(const Callable &p_callable) {
...
 	do_op.callable = p_callable;
-	do_op.object = p_callable.get_object_id();
+	do_op.object = object_id;
 	if (Object::cast_to<RefCounted>(object)) {
 		do_op.ref = Ref<RefCounted>(Object::cast_to<RefCounted>(object));
 	}
 	do_op.type = Operation::TYPE_METHOD;
 	do_op.name = p_callable.get_method();
+	if (do_op.name == StringName()) {
+		// There's no `get_method` for custom callables, so use `operator String` instead
```
Comment (KoBeWi) [inline core/object/undo_redo.cpp:152]:
```suggestion
		// There's no `get_method()` for custom callables, so use `operator String` instead.
```

File: servers/physics_2d/godot_area_2d.cpp [hunk 9]
```diff
@@ -78,13 +78,6 @@ void GodotArea2D::set_space(GodotSpace2D *p_space) {
 }
 
 void GodotArea2D::set_monitor_callback(const Callable &p_callback) {
-	ObjectID id = p_callback.get_object_id();
-
-	if (id == monitor_callback.get_object_id()) {
```
Comment (KoBeWi) [inline servers/physics_2d/godot_area_2d.cpp:83]: This looks like some optimization and should work without an object too (though the condition probably needs changing). Why was this removed?

File: core/object/object.cpp [hunk 10]
```diff
@@ -1313,8 +1313,14 @@ void Object::get_signals_connected_to_this(List<Connection> *p_connections) cons
 Error Object::connect(const StringName &p_signal, const Callable &p_callable, uint32_t p_flags) {
 	ERR_FAIL_COND_V_MSG(p_callable.is_null(), ERR_INVALID_PARAMETER, "Cannot connect to '" + p_signal + "': the provided callable is null.");
 
-	Object *target_object = p_callable.get_object();
-	ERR_FAIL_NULL_V_MSG(target_object, ERR_INVALID_PARAMETER, "Cannot connect to '" + p_signal + "' to callable '" + p_callable + "': the callable object is null.");
+	if (p_callable.is_standard()) {
+		// NOTE: This branch should probably removed in favor of the `is_valid` branch, but there exist some classes
```
Comment (KoBeWi) [inline core/object/object.cpp:1317]:
```suggestion
		// FIXME: This branch should probably removed in favor of the `is_valid` branch, but there exist some classes
```
Or TODO.
NOTE doesn't signify that it's something actionable.

File: servers/physics_2d/godot_area_2d.cpp (hunk 9 above)
Comment (maiself) [reply servers/physics_2d/godot_area_2d.cpp:83]: For object-less callables this will always fall into the optimized path because `get_object_id()` will always return `ObjectID(0)`, so for those callables its no good, and why I removed it. Im not sure what this condition could be changed to.

I tried to see what this optimization and the code that follows was about, but I'm not so familiar with the physics servers inner workings. This code goes all they way back to the original opensource commit, so no explanation there. When I tried to follow all the paths that could lead here it seemed that the optimized path doesn't ever get used in the current code, tho I didn't test it.

In short, I don't know why that code is there in the first place, how to properly change it to support object-less callables (tho maybe just adding `p_callback.is_standard() && ` to the condition would be enough?), and it doesn't appear that removing it would harm anything.

File: core/object/object.cpp (hunk 10 above)
Comment (maiself) [reply core/object/object.cpp:1317]: Ah, I like the idea of using something that marks this as actionable, will change!

File: servers/physics_2d/godot_area_2d.cpp (hunk 9 above)
Comment (KoBeWi) [reply servers/physics_2d/godot_area_2d.cpp:83]: Maybe something like `id.is_valid() && id == monitor_callback.get_object_id()`?
This code allows to swap callbacks without clearing shapes if the object is the same. So the condition I suggested would make it skipped if the new callback has no object.

That said, I'm also not very familiar with physics code either. This "optimization" doesn't look very important; normally the callbacks are changed only when toggling area's `monitoring` property, which is rather uncommon thing to do.

**Review by KoBeWi — APPROVED** [review]
I only tested Tweens, but the code seems fine.

**Comment by akien-mga** [comment]
Thanks!
